### PR TITLE
Support app creation in `goal app method`

### LINKED ISSUE #3332: Support app creation in `goal app method`
## Summary

When `goal app method` was originally added we decided to not support app creation transactions because it was undecided how the ABI would work with them. Now, https://github.com/algorandfoundation/ARCs/pull/57 has made it clear ABI method calls should work on creation as well.

## Scope

Allow `goal app method` to be able to create apps. To do this, we need to:
* Somehow indicate an app creation. Initial idea: this could be done with specifying 0 for the app ID, or with an explicit `--create` flag.
* Accept the approval program, clear program, schema counts, and extra pages value for the creation transaction
* Somehow return the created app ID in addition to the method return value

### PR DESCRIPTION
**Pull request by jasonpaulos**
## Summary

Supports app creation in `goal app method`. To create an app, the `--create` flag must be passed instead of the `--app-id` flag. Additionally flags for schema sizes, extra pages, and programs should be passed. After the transaction is confirmed, `goal` will print out the created app ID in addition to the ABI return value.

Here's an example:

```bash
$ goal app method --method "create(uint64)uint64" --arg "1234" --create --approval-prog $APPROVALPROG --clear-prog $CLEARPROG --global-byteslices 0 --global-ints 0 --local-byteslices 1 --local-ints 0 --extra-pages 0 --from WF3PXLVCDHILATGW6P7BOLC7TJE7TOLZM2BQCS24KKX2B4FGMG5Y2NDHWA
Attempting to create app (approval size 500, hash EZD2SJS7UJI45CGG4GEVMCTPZCJ7GQNDUG74GU74C6WKOD2TEFAA; clear size 5, hash YOE6C22GHCTKAN3HU4SE5PGIPN5UKXAJTXCQUPJ3KKF5HOAH646A)
Issued 1 transaction(s):
Issued transaction from account WF3PXLVCDHILATGW6P7BOLC7TJE7TOLZM2BQCS24KKX2B4FGMG5Y2NDHWA, txid 2LOUCI2XOG6XDC6JZWPTVPWHPTF53JSABIEZDBBZGRWII4VM752Q (fee 1000)
Transaction 2LOUCI2XOG6XDC6JZWPTVPWHPTF53JSABIEZDBBZGRWII4VM752Q still pending as of round 2
Transaction 2LOUCI2XOG6XDC6JZWPTVPWHPTF53JSABIEZDBBZGRWII4VM752Q still pending as of round 3
Transaction 2LOUCI2XOG6XDC6JZWPTVPWHPTF53JSABIEZDBBZGRWII4VM752Q committed in round 4
Created app with app index 2
method create(uint64)uint64 succeeded with output: 2468
```

Closes #3332 

## Test Plan

Added e2e test

### CODE REVIEW

File: test/scripts/e2e_subs/tealprogs/app-abi-method-example.teal
Comment (jasonpaulos) [inline test/scripts/e2e_subs/tealprogs/app-abi-method-example.teal:1]: The PyTeal diff is here: https://gist.github.com/jasonpaulos/99e4f8a75f2fc2ec9b8073c064530359/revisions#diff-7f2d296a873a322392d91f69eebfc6993a857884a7e55cef9af800a408ef378f

File: cmd/goal/application.go
Comment (ahangsu) [inline cmd/goal/application.go:1413]: minor question: do we need to concern the existence of `*resp.ApplicationIndex` on the condition of:
- `client.PendingTransactionInformationV2` successfully responded with no error
- creating an application from app-call transaction

If not, it seems okay to lift the check of null-ptr?

File: test/scripts/e2e_subs/e2e-app-abi-method.sh
Comment (ahangsu) [inline test/scripts/e2e_subs/e2e-app-abi-method.sh:20]: minor question: why we want 2 versions of approval-prog, but it seems to me that the behavior is the same?

Comment (jasonpaulos) [reply test/scripts/e2e_subs/e2e-app-abi-method.sh:20]: `simple-v2.teal` is the clear state program used for creation, and then we update the app to use `simple-v3.teal` later on.

File: cmd/goal/application.go
Comment (jasonpaulos) [reply cmd/goal/application.go:1413]: I'd rather have the specific checks in an if statement than assume the pointer will always be valid, since you never know what could change in the future.

For what it's worth, the existing create app command has a similar check too: https://github.com/algorand/go-algorand/blob/2e09ebfca1ca46231e43cd62c48a1025ce7c0a11/cmd/goal/application.go#L490-L492

File: test/scripts/e2e_subs/e2e-app-abi-method.sh
Comment (ahangsu) [reply test/scripts/e2e_subs/e2e-app-abi-method.sh:20]: That makes sense to me.

File: test/scripts/e2e_subs/e2e-app-abi-method.sh
Comment (ahangsu) [inline test/scripts/e2e_subs/e2e-app-abi-method.sh:26]: the error message seems off to me: `optIn(string)string`?

File: test/scripts/e2e_subs/e2e-app-abi-method.sh
Comment (ahangsu) [inline test/scripts/e2e_subs/e2e-app-abi-method.sh:102]: same here for the error message.

File: test/scripts/e2e_subs/e2e-app-abi-method.sh
Comment (jasonpaulos) [reply test/scripts/e2e_subs/e2e-app-abi-method.sh:26]: Yes, I'll fix it

**Review by ahangsu — COMMENTED** [review]
Look good to me, waiting on @jannotti to review

File: cmd/goal/application.go
Comment (jannotti) [inline cmd/goal/application.go:51]: This should really be "appCreation" or "contractCreation", right?

File: cmd/goal/application.go
Comment (jannotti) [inline cmd/goal/application.go:1195]: This is obviously a decision that was made before, this PR doesn't change it.  But do you think we should really be requiring on-complete to be set always?  Rather than default to no-op?

Comment (jasonpaulos) [reply cmd/goal/application.go:1195]: Actually you don't need to explicitly pass it in, it will default to "NoOp": https://github.com/algorand/go-algorand/blob/1642a3214a12cd49843d62b2e18a1d3e516e7316/cmd/goal/application.go#L125

The e2e test for creation does not pass in `--on-completion`.

File: cmd/goal/application.go
Comment (jasonpaulos) [reply cmd/goal/application.go:51]: Well I meant it to mean creation for the method subcommand

Comment (ahangsu) [reply cmd/goal/application.go:51]: `methodAppCreation`? (it's always hard to find a good name)

File: cmd/goal/application.go
Comment (jannotti) [reply cmd/goal/application.go:1195]: oh, yeah. I guess I wish there was no "must" on that method name then.  While it's true, it seems like it's trying to convey something important that never comes up in practice because of the default. But I don't care _that_ much.

File: cmd/goal/application.go
Comment (jannotti) [reply cmd/goal/application.go:51]: My objection is that when you read any code that uses it, it reads wrong.  It'll say, essentially
```
if methodCreation {
 // create the app
}
```
As with the other comment, whatever you want to go with is fine.  Let me know when to merge.

Comment (jasonpaulos) [reply cmd/goal/application.go:51]: Good points, I'll rename to `methodCreatesApp`

File: cmd/goal/application.go
Comment (jasonpaulos) [reply cmd/goal/application.go:1195]: I agree it's a bit misleading, but I'll leave it for now since there are multiple other helper methods in the goal code that use "must" to convey they will either succeed or panic.

**Review by jannotti — APPROVED** [review]

**Pull request merged**
